File: src/test/java/org/railwaystations/rsapi/adapter/in/task/WebDavSyncTaskTest.java
package org.railwaystations.rsapi.adapter.in.task;

import com.github.tomakehurst.wiremock.client.BasicCredentials;
import com.github.tomakehurst.wiremock.http.RequestMethod;
import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo;
import com.github.tomakehurst.wiremock.junit5.WireMockTest;
import com.github.tomakehurst.wiremock.matching.RequestPatternBuilder;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.railwaystations.rsapi.adapter.out.db.InboxDao;
import org.railwaystations.rsapi.core.model.Coordinates;
import org.railwaystations.rsapi.core.model.InboxEntry;
import org.railwaystations.rsapi.core.ports.out.PhotoStorage;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.util.List;

import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
import static com.github.tomakehurst.wiremock.client.WireMock.badRequest;
import static com.github.tomakehurst.wiremock.client.WireMock.binaryEqualTo;
import static com.github.tomakehurst.wiremock.client.WireMock.created;
import static com.github.tomakehurst.wiremock.client.WireMock.delete;
import static com.github.tomakehurst.wiremock.client.WireMock.deleteRequestedFor;
import static com.github.tomakehurst.wiremock.client.WireMock.get;
import static com.github.tomakehurst.wiremock.client.WireMock.getRequestedFor;
import static com.github.tomakehurst.wiremock.client.WireMock.noContent;
import static com.github.tomakehurst.wiremock.client.WireMock.ok;
import static com.github.tomakehurst.wiremock.client.WireMock.put;
import static com.github.tomakehurst.wiremock.client.WireMock.putRequestedFor;
import static com.github.tomakehurst.wiremock.client.WireMock.request;
import static com.github.tomakehurst.wiremock.client.WireMock.serverError;
import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
import static com.github.tomakehurst.wiremock.client.WireMock.unauthorized;
import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
import static com.github.tomakehurst.wiremock.client.WireMock.verify;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@WireMockTest
class WebDavSyncTaskTest {

    static final String FILENAME = "1.jpg";
    static final String TO_PROCESS_PATH = "/toProcessPath";
    static final String TO_PROCESS_PATH_FILE = TO_PROCESS_PATH + "/" + FILENAME;
    static final String PROCESSED_PATH = "/processedPath";
    static final String PROCESSED_PATH_FILE = PROCESSED_PATH + "/" + FILENAME;
    static final String USERNAME = "username";
    static final String PASSWORD = "password";
    public static final String PROPFIND_METHOD = "PROPFIND";

    Path tempdir;
    PhotoStorage photoStorage;
    InboxDao inboxDao;

    @BeforeEach
    void setup() throws IOException {
        tempdir = Files.createTempDirectory("rsapi");
        photoStorage = mock(PhotoStorage.class);
        inboxDao = mock(InboxDao.class);
        when(inboxDao.findPendingInboxEntries()).thenReturn(List.of(createInboxEntry()));
    }

    @Test
    void shouldAuthenticate(final WireMockRuntimeInfo wmRuntimeInfo) throws IOException {
        final var task = createWebDavSyncTask(wmRuntimeInfo);
        final var toProcessPath = createFile(tempdir);
        when(photoStorage.getInboxToProcessFile(FILENAME)).thenReturn(toProcessPath);
        when(photoStorage.getInboxProcessedFile(FILENAME)).thenReturn(tempdir.resolve(FILENAME));

        stubFor(put(TO_PROCESS_PATH_FILE)
                .willReturn(unauthorized().withHeader("WWW-Authenticate", "Basic realm=\"realm\"")));
        stubFor(put(TO_PROCESS_PATH_FILE)
                .withBasicAuth(USERNAME, PASSWORD)
                .willReturn(created()));
        stubPropfindEmpty();

        task.syncWebDav();

        verify(putRequestedFor(urlEqualTo(TO_PROCESS_PATH_FILE))
                .withRequestBody(binaryEqualTo(FILENAME.getBytes(StandardCharsets.UTF_8)
                )));
        verify(putRequestedFor(urlEqualTo(TO_PROCESS_PATH_FILE))
                .withBasicAuth(new BasicCredentials(USERNAME, PASSWORD))
                .withRequestBody(binaryEqualTo(FILENAME.getBytes(StandardCharsets.UTF_8)
                )));
    }

    @Test
    void shouldUploadToProcessFile(final WireMockRuntimeInfo wmRuntimeInfo) throws IOException {
        final var task = createWebDavSyncTask(wmRuntimeInfo);
        final var toProcessPath1 = createFile(tempdir);
        when(photoStorage.getInboxToProcessFile(FILENAME)).thenReturn(toProcessPath1);
        when(photoStorage.getInboxProcessedFile(FILENAME)).thenReturn(tempdir.resolve(FILENAME));

        stubFor(put(TO_PROCESS_PATH_FILE).willReturn(created()));
        stubPropfindEmpty();

        task.syncWebDav();

        verify(putRequestedFor(urlEqualTo(TO_PROCESS_PATH_FILE))
                .withRequestBody(binaryEqualTo(FILENAME.getBytes(StandardCharsets.UTF_8))));

        assertThat(Files.exists(toProcessPath1)).isFalse();
    }

    @Test
    void shouldKeepFileIfUploadFails(final WireMockRuntimeInfo wmRuntimeInfo) throws IOException {
        final var task = createWebDavSyncTask(wmRuntimeInfo);
        final var toProcessPath = createFile(tempdir);
        when(photoStorage.getInboxToProcessFile(FILENAME)).thenReturn(toProcessPath);
        when(photoStorage.getInboxProcessedFile(FILENAME)).thenReturn(tempdir.resolve(FILENAME));

        stubFor(put(TO_PROCESS_PATH_FILE).willReturn(badRequest()));
        stubPropfindEmpty();

        task.syncWebDav();

        verify(putRequestedFor(urlEqualTo(TO_PROCESS_PATH_FILE))
                .withRequestBody(binaryEqualTo(FILENAME.getBytes(StandardCharsets.UTF_8)
                )));

        assertThat(Files.exists(toProcessPath)).isTrue();
    }

    @Test
    void shouldDownloadProcessedFile(final WireMockRuntimeInfo wmRuntimeInfo) {
        final var task = createWebDavSyncTask(wmRuntimeInfo);
        when(photoStorage.getInboxToProcessFile(FILENAME)).thenReturn(tempdir.resolve(FILENAME));
        final var processedPath = tempdir.resolve(FILENAME);
        when(photoStorage.getInboxProcessedFile(FILENAME)).thenReturn(processedPath);

        stubPropfind(createMultistatusResponseForFile());
        stubFor(get(PROCESSED_PATH_FILE).willReturn(ok(FILENAME)));
        stubFor(delete(PROCESSED_PATH_FILE).willReturn(noContent()));

        task.syncWebDav();

        verifyPropfindRequest();
        verify(getRequestedFor(urlEqualTo(PROCESSED_PATH_FILE)));
        verify(deleteRequestedFor(urlEqualTo(PROCESSED_PATH_FILE)));

        assertThat(Files.exists(processedPath)).isTrue();
    }

    private void verifyPropfindRequest() {
        verify(new RequestPatternBuilder(RequestMethod.fromString(PROPFIND_METHOD), urlEqualTo(PROCESSED_PATH)));
    }

    private void stubPropfindEmpty() {
        stubPropfind("");
    }

    private void stubPropfind(final String multistatusResponse) {
        stubFor(request(PROPFIND_METHOD, urlPathEqualTo(PROCESSED_PATH)).willReturn(
                aResponse().withStatus(207).withBody("""
                        <?xml version="1.0"?>
                        <d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns"
                                       xmlns:nc="http://nextcloud.org/ns">
                            <d:response>
                                <d:href>%s/</d:href>
                                <d:propstat>
                                    <d:prop>
                                        <d:resourcetype>
                                            <d:collection/>
                                        </d:resourcetype>
                                    </d:prop>
                                    <d:status>HTTP/1.1 200 OK</d:status>
                                </d:propstat>
                            </d:response>
                            <d:response>
                                <d:href>%s/Readme.md</d:href>
                                <d:propstat>
                                    <d:prop>
                                        <d:resourcetype/>
                                    </d:prop>
                                    <d:status>HTTP/1.1 200 OK</d:status>
                                </d:propstat>
                            </d:response>
                            %s
                        </d:multistatus>
                        """.formatted(PROCESSED_PATH, PROCESSED_PATH, multistatusResponse))));
    }

    String createMultistatusResponseForFile() {
        return """
                <d:response>
                    <d:href>%s</d:href>
                    <d:propstat>
                        <d:prop>
                            <d:resourcetype/>
                        </d:prop>
                        <d:status>HTTP/1.1 200 OK</d:status>
                    </d:propstat>
                </d:response>
                """.formatted(PROCESSED_PATH_FILE);
    }

    @Test
    void shouldNotSendDeleteIfDownloadFailed(final WireMockRuntimeInfo wmRuntimeInfo) {
        final var task = createWebDavSyncTask(wmRuntimeInfo);
        when(photoStorage.getInboxToProcessFile(FILENAME)).thenReturn(tempdir.resolve(FILENAME));
        final var processedPath = tempdir.resolve(FILENAME);
        when(photoStorage.getInboxProcessedFile(FILENAME)).thenReturn(processedPath);

        stubPropfind(createMultistatusResponseForFile());
        stubFor(get(PROCESSED_PATH_FILE).willReturn(serverError()));
        stubFor(delete(PROCESSED_PATH_FILE).willReturn(noContent()));

        task.syncWebDav();

        verifyPropfindRequest();
        verify(getRequestedFor(urlEqualTo(PROCESSED_PATH_FILE)));
        verify(0, deleteRequestedFor(urlEqualTo(PROCESSED_PATH_FILE)));

        assertThat(Files.exists(processedPath)).isFalse();
    }

    @NotNull
    WebDavSyncTask createWebDavSyncTask(final WireMockRuntimeInfo wmRuntimeInfo) {
        final var config = new WebDavSyncConfig(wmRuntimeInfo.getHttpBaseUrl() + TO_PROCESS_PATH,
                wmRuntimeInfo.getHttpBaseUrl() + PROCESSED_PATH, USERNAME, PASSWORD);
        return new WebDavSyncTask(config, photoStorage, inboxDao);
    }

    @NotNull
    InboxEntry createInboxEntry() {
        return InboxEntry.builder()
                .id(1)
                .countryCode("de")
                .stationId("stationId")
                .title("title")
                .coordinates(new Coordinates())
                .photographerId(1)
                .photographerNickname("nickname")
                .extension("jpg")
                .createdAt(Instant.now())
                .build();
    }

    @NotNull
    Path createFile(final Path dir) throws IOException {
        final var path = dir.resolve(WebDavSyncTaskTest.FILENAME);
        Files.writeString(path, WebDavSyncTaskTest.FILENAME);
        return path;
    }

}

File: adapter/src/test/kotlin/org/railwaystations/rsapi/adapter/task/WebDavSyncTaskTest.kt
package org.railwaystations.rsapi.adapter.task

import com.github.tomakehurst.wiremock.client.BasicCredentials
import com.github.tomakehurst.wiremock.client.WireMock.aResponse
import com.github.tomakehurst.wiremock.client.WireMock.badRequest
import com.github.tomakehurst.wiremock.client.WireMock.binaryEqualTo
import com.github.tomakehurst.wiremock.client.WireMock.created
import com.github.tomakehurst.wiremock.client.WireMock.delete
import com.github.tomakehurst.wiremock.client.WireMock.deleteRequestedFor
import com.github.tomakehurst.wiremock.client.WireMock.get
import com.github.tomakehurst.wiremock.client.WireMock.getRequestedFor
import com.github.tomakehurst.wiremock.client.WireMock.noContent
import com.github.tomakehurst.wiremock.client.WireMock.ok
import com.github.tomakehurst.wiremock.client.WireMock.put
import com.github.tomakehurst.wiremock.client.WireMock.putRequestedFor
import com.github.tomakehurst.wiremock.client.WireMock.request
import com.github.tomakehurst.wiremock.client.WireMock.serverError
import com.github.tomakehurst.wiremock.client.WireMock.stubFor
import com.github.tomakehurst.wiremock.client.WireMock.unauthorized
import com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo
import com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo
import com.github.tomakehurst.wiremock.client.WireMock.verify
import com.github.tomakehurst.wiremock.http.RequestMethod
import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo
import com.github.tomakehurst.wiremock.junit5.WireMockTest
import com.github.tomakehurst.wiremock.matching.RequestPatternBuilder
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.railwaystations.rsapi.adapter.db.InboxDao
import org.railwaystations.rsapi.core.model.InboxEntry
import org.railwaystations.rsapi.core.ports.outbound.PhotoStoragePort
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

const val FILENAME: String = "1.jpg"
const val TO_PROCESS_PATH: String = "/toProcessPath"
const val TO_PROCESS_PATH_FILE: String = "$TO_PROCESS_PATH/$FILENAME"
const val PROCESSED_PATH: String = "/processedPath"
const val PROCESSED_PATH_FILE: String = "$PROCESSED_PATH/$FILENAME"
const val USERNAME: String = "username"
const val PASSWORD: String = "password"
const val PROPFIND_METHOD: String = "PROPFIND"

@WireMockTest
internal class WebDavSyncTaskTest {
    private lateinit var tempdir: Path
    private lateinit var photoStoragePort: PhotoStoragePort
    private lateinit var inboxDao: InboxDao

    @BeforeEach
    fun setup() {
        tempdir = Files.createTempDirectory("rsapi")
        photoStoragePort = mockk<PhotoStoragePort>()
        inboxDao = mockk<InboxDao>()
        every { inboxDao.findPendingInboxEntries() } returns listOf(createInboxEntry())
    }

    @Test
    fun shouldAuthenticate(wmRuntimeInfo: WireMockRuntimeInfo) {
        val task = createWebDavSyncTask(wmRuntimeInfo)
        val toProcessPath = createFile(tempdir)
        every { photoStoragePort.getInboxToProcessFile(FILENAME) } returns toProcessPath
        every { photoStoragePort.getInboxProcessedFile(FILENAME) } returns tempdir.resolve(FILENAME)

        stubFor(
            put(TO_PROCESS_PATH_FILE)
                .willReturn(unauthorized().withHeader("WWW-Authenticate", "Basic realm=\"realm\""))
        )
        stubFor(
            put(TO_PROCESS_PATH_FILE)
                .withBasicAuth(USERNAME, PASSWORD)
                .willReturn(created())
        )
        stubPropfindEmpty()

        task.syncWebDav()

        verify(
            putRequestedFor(urlEqualTo(TO_PROCESS_PATH_FILE))
                .withRequestBody(
                    binaryEqualTo(
                        FILENAME.toByteArray(StandardCharsets.UTF_8)
                    )
                )
        )
        verify(
            putRequestedFor(urlEqualTo(TO_PROCESS_PATH_FILE))
                .withBasicAuth(BasicCredentials(USERNAME, PASSWORD))
                .withRequestBody(
                    binaryEqualTo(
                        FILENAME.toByteArray(StandardCharsets.UTF_8)
                    )
                )
        )
    }

    @Test
    fun shouldUploadToProcessFile(wmRuntimeInfo: WireMockRuntimeInfo) {
        val task = createWebDavSyncTask(wmRuntimeInfo)
        val toProcessPath1 = createFile(tempdir)
        every { photoStoragePort.getInboxToProcessFile(FILENAME) } returns toProcessPath1
        every { photoStoragePort.getInboxProcessedFile(FILENAME) } returns tempdir.resolve(FILENAME)

        stubFor(put(TO_PROCESS_PATH_FILE).willReturn(created()))
        stubPropfindEmpty()

        task.syncWebDav()

        verify(
            putRequestedFor(urlEqualTo(TO_PROCESS_PATH_FILE))
                .withRequestBody(binaryEqualTo(FILENAME.toByteArray(StandardCharsets.UTF_8)))
        )

        Assertions.assertThat(Files.exists(toProcessPath1)).isFalse()
    }

    @Test
    fun shouldKeepFileIfUploadFails(wmRuntimeInfo: WireMockRuntimeInfo) {
        val task = createWebDavSyncTask(wmRuntimeInfo)
        val toProcessPath = createFile(tempdir)
        every { photoStoragePort.getInboxToProcessFile(FILENAME) } returns toProcessPath
        every { photoStoragePort.getInboxProcessedFile(FILENAME) } returns tempdir.resolve(FILENAME)

        stubFor(put(TO_PROCESS_PATH_FILE).willReturn(badRequest()))
        stubPropfindEmpty()

        task.syncWebDav()

        verify(
            putRequestedFor(urlEqualTo(TO_PROCESS_PATH_FILE))
                .withRequestBody(
                    binaryEqualTo(
                        FILENAME.toByteArray(StandardCharsets.UTF_8)
                    )
                )
        )

        Assertions.assertThat(Files.exists(toProcessPath)).isTrue()
    }

    @Test
    fun shouldDownloadProcessedFile(wmRuntimeInfo: WireMockRuntimeInfo) {
        val task = createWebDavSyncTask(wmRuntimeInfo)
        every { photoStoragePort.getInboxToProcessFile(FILENAME) } returns tempdir.resolve(FILENAME)
        val processedPath = tempdir.resolve(FILENAME)
        every { photoStoragePort.getInboxProcessedFile(FILENAME) } returns processedPath

        stubPropfind(createMultistatusResponseForFile())
        stubFor(get(PROCESSED_PATH_FILE).willReturn(ok(FILENAME)))
        stubFor(delete(PROCESSED_PATH_FILE).willReturn(noContent()))

        task.syncWebDav()

        verifyPropfindRequest()
        verify(getRequestedFor(urlEqualTo(PROCESSED_PATH_FILE)))
        verify(deleteRequestedFor(urlEqualTo(PROCESSED_PATH_FILE)))

        Assertions.assertThat(Files.exists(processedPath)).isTrue()
    }

    private fun verifyPropfindRequest() {
        verify(
            RequestPatternBuilder(
                RequestMethod.fromString(PROPFIND_METHOD), urlEqualTo(
                    PROCESSED_PATH
                )
            )
        )
    }

    private fun stubPropfindEmpty() {
        stubPropfind("")
    }

    private fun stubPropfind(multistatusResponse: String) {
        stubFor(
            request(PROPFIND_METHOD, urlPathEqualTo(PROCESSED_PATH)).willReturn(
                aResponse().withStatus(207).withBody(
                    """
                        <?xml version="1.0"?>
                        <d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns"
                                       xmlns:nc="http://nextcloud.org/ns">
                            <d:response>
                                <d:href>%s/</d:href>
                                <d:propstat>
                                    <d:prop>
                                        <d:resourcetype>
                                            <d:collection/>
                                        </d:resourcetype>
                                    </d:prop>
                                    <d:status>HTTP/1.1 200 OK</d:status>
                                </d:propstat>
                            </d:response>
                            <d:response>
                                <d:href>%s/Readme.md</d:href>
                                <d:propstat>
                                    <d:prop>
                                        <d:resourcetype/>
                                    </d:prop>
                                    <d:status>HTTP/1.1 200 OK</d:status>
                                </d:propstat>
                            </d:response>
                            %s
                        </d:multistatus>
                        
                        """.trimIndent().format(PROCESSED_PATH, PROCESSED_PATH, multistatusResponse)
                )
            )
        )
    }

    private fun createMultistatusResponseForFile(): String {
        return """
                <d:response>
                    <d:href>%s</d:href>
                    <d:propstat>
                        <d:prop>
                            <d:resourcetype/>
                        </d:prop>
                        <d:status>HTTP/1.1 200 OK</d:status>
                    </d:propstat>
                </d:response>
                
                """.trimIndent().format(PROCESSED_PATH_FILE)
    }

    @Test
    fun shouldNotSendDeleteIfDownloadFailed(wmRuntimeInfo: WireMockRuntimeInfo) {
        val task = createWebDavSyncTask(wmRuntimeInfo)
        every { photoStoragePort.getInboxToProcessFile(FILENAME) } returns tempdir.resolve(FILENAME)
        val processedPath = tempdir.resolve(FILENAME)
        every { photoStoragePort.getInboxProcessedFile(FILENAME) } returns processedPath

        stubPropfind(createMultistatusResponseForFile())
        stubFor(get(PROCESSED_PATH_FILE).willReturn(serverError()))
        stubFor(delete(PROCESSED_PATH_FILE).willReturn(noContent()))

        task.syncWebDav()

        verifyPropfindRequest()
        verify(getRequestedFor(urlEqualTo(PROCESSED_PATH_FILE)))
        verify(0, deleteRequestedFor(urlEqualTo(PROCESSED_PATH_FILE)))

        Assertions.assertThat(Files.exists(processedPath)).isFalse()
    }

    private fun createWebDavSyncTask(wmRuntimeInfo: WireMockRuntimeInfo): WebDavSyncTask {
        val config = WebDavSyncConfig(
            toProcessUrl = wmRuntimeInfo.httpBaseUrl + TO_PROCESS_PATH,
            processedUrl = wmRuntimeInfo.httpBaseUrl + PROCESSED_PATH, user = USERNAME, password = PASSWORD
        )
        return WebDavSyncTask(config, photoStoragePort, inboxDao)
    }

    private fun createInboxEntry(): InboxEntry {
        return InboxEntry(
            id = 1,
            countryCode = "de",
            stationId = "stationId",
            title = "title",
            photographerNickname = "nickname",
            extension = "jpg",
        )
    }

    private fun createFile(dir: Path?): Path {
        val path = dir!!.resolve(FILENAME)
        Files.writeString(path, FILENAME)
        return path
    }

}